### DISCUSSION TITLE
What Mechanism does littlenavmap use to connect to Simulators?

### DISCUSSION THREAD

**Discussion by sosukeinu** [albar965/littlenavmap#1104] [General]
Hello,
I'm trying to judge the amount of work needed to add a connection to a different simulator. What is the connection mechanism/protocol for connecting to simulators? Is this a one-way connection, or two-way? Does littlenavmap use DIS? I'm trying to understand if I might use this software with something like AFSIM. Thank you for your time.

**Comment by albar965** [comment]
LNM uses SimConnect to connect to MSFS, FSX and P3D (code is here: https://github.com/albar965/atools/tree/release/4.0/src/fs/sc). This is more or less bi-directional if you consider weather requests. I suppose this is not usable to adapt to other sims.

X-Plane needs a plugin to connect to Little Navmap: https://github.com/albar965/littlexpconnect/tree/release/1.2
LNM communicates using shared memory with this plugin. This is uni-directional but I have plans to change this.

Little Navconnect connects to the simulators the same as LNM but provides the data across a plain TCP/IP connection. Have plans to replace this with RestAPI.

In all of the above cases the class SimConnectData (https://github.com/albar965/atools/blob/release/4.0/src/fs/sc/simconnectdata.h) and its aggregated are used to transfer the data. The content of objects is directly serialized into and from a binary stream using Qt streams. The stream data is platform independent.

Additionally there is a web API: https://www.littlenavmap.org/manuals/littlenavmap/release/latest/en/WEBSERVER.html#webapi
which allows to fetch data using a RestAPI and JSON. This is not complete and uni-directional.

No, I don't use DIS (supposedly this one https://open-dis.org/ ?).

Alex

**Reply by sosukeinu** [reply]
Thank you for the fast response. I might be able to use the web API for now,...I'll poke around at it, and see what I can use. By the way, I would be remiss not to say,...thank you so much for such an great piece of software! It is very impressive, and quite helpful. Have you ever considered something like gRPC using Protocol Buffers as a way of making that same web API a two-way affair?

**Reply by albar965** [reply]
Thank you for the kind words!

> Have you ever considered something like gRPC using Protocol Buffers as a way of making that same web API a two-way affair?

Always but time is limited. gRPC and Protocol Buffers are really tempting since HTTP is probably too slow to transfer data from sim to program. I need to rewrite the whole networking stuff anyway.
